spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/actuate/AbstractGatewayControllerEndpoint.java

```diff
@@ -93,11 +97,28 @@ public void setApplicationEventPublisher(ApplicationEventPublisher publisher) {
 	// TODO: Add uncommited or new but not active routes endpoint
 
 	@PostMapping("/refresh")
-	public Mono<Void> refresh() {
-		this.publisher.publishEvent(new RefreshRoutesEvent(this));
+	public Mono<Void> refresh(@RequestParam(value = "metadata", required = false) List<String> byMetadata) {
+		publishRefreshEvent(byMetadata);
 		return Mono.empty();
 	}
 
+	private void publishRefreshEvent(List<String> byMetadata) {
+		RefreshRoutesEvent event;
+		if (!CollectionUtils.isEmpty(byMetadata)) {
+			Map<String, String> scopeMetadata = convertToMap(byMetadata);
+			event = new ScopedRefreshRoutesEvent(this, scopeMetadata);
+		}
+		else {
+			event = new RefreshRoutesEvent(this);
+		}
+		this.publisher.publishEvent(event);
+	}
+
+	private Map<String, String> convertToMap(List<String> byMetadata) {
+		return byMetadata.stream().map(keyValue -> keyValue.split(":"))
```

Comment:
Should we filter out the pairs that have no value (using `StringUtils.hasLength()`) instead of inserting a null value?